File: 2015/15/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const REGEX = /^[a-zA-Z]+: capacity (-?\d+), durability (-?\d+), flavor (-?\d+), texture (-?\d+), calories (-?\d+)$/;
const MAX = 100;

// --- Functions ---
function notNegative(value) {
	if (value < 0) return 0;
	return value;
}

// --- Variables ---
let ingredients = [];
let max1 = Number.MIN_VALUE;
let max2 = Number.MIN_VALUE;

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	const cap = line.match(REGEX);
	const capacity = Number(cap[1]);
	const durability = Number(cap[2]);
	const flavor = Number(cap[3]);
	const texture = Number(cap[4]);
	const calories = Number(cap[5]);
	ingredients.push({capacity, durability, flavor, texture, calories});
});
await new Promise(res => reader.once("close", res));

// --- Calculate combinations ---
let list = Array.from({length: ingredients.length - 1}, () => 0);
list.push(MAX);

let last = list.length - 1;
outer: while (list[0] < MAX) {
	let capacity = 0;
	let durability = 0;
	let flavor = 0;
	let texture = 0;
	let calories = 0;
	for (let i = 0; i < ingredients.length; i++) {
		capacity += list[i] * ingredients[i].capacity;
		durability += list[i] * ingredients[i].durability;
		flavor += list[i] * ingredients[i].flavor;
		texture += list[i] * ingredients[i].texture;
		calories += list[i] * ingredients[i].calories;
	}
	capacity = notNegative(capacity);
	durability = notNegative(durability);
	flavor = notNegative(flavor);
	texture = notNegative(texture);
	let total = capacity * durability * flavor * texture;
	max1 = Math.max(max1, total);
	if (calories === 500) max2 = Math.max(max2, total);

	// --- Calcualte next recipe ---
	let i = last - 1;
	let carry = true;
	while (carry) {
		let limit = MAX;
		if (i > 0) for (let j = 0; j < i; j++) limit -= list[i];
		list[i] += 1;
		if (list[i] > limit) {
			list[i] = 0;
			if (i > 0) { i--; } else { break  outer; }
		} else {
			carry = false;
		}
		list[last] = MAX;
		for (let j = 0; j < last; j++) list[last] -= list[j];
	}
}

// --- Puzzle 1 ---
console.log("1. Total score of the highest-scoring cookie: ", max1.toLocaleString());

// --- Puzzle 2 ---
console.log("2. Total score of the highest-scoring cookie: ", max2.toLocaleString());

File: 2015/15/python/main.py
import re

# --- Constants --
INPUT_FILE = "../input/input.txt"
REGEX = r"[a-zA-Z]+: capacity (-?\d+), durability (-?\d+), flavor (-?\d+), texture (-?\d+), calories (-?\d+)$";
MAX = 100;

# --- Functions ---
def not_negative(value):
    if value < 0: return 0
    return value

# --- Variables ---
ingredients = []
max1 = float("-inf")
max2 = float("-inf")

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
while True:
    line = file.readline()
    if not line: break
    line = line.strip()

    cap = re.search(REGEX, line)
    capacity   = int(cap[1])
    durability = int(cap[2])
    flavor     = int(cap[3])
    texture    = int(cap[4])
    calories   = int(cap[5])
    ingredients.append((capacity, durability, flavor, texture, calories))
file.close()

# --- Calculate combinations ---
last = len(ingredients) - 1
ls = [0] * last
ls.append(MAX)

is_done = False
while ls[0] < MAX and not is_done:
    capacity = 0
    durability = 0
    flavor = 0
    texture = 0
    calories = 0
    for i in range(len(ingredients)):
        capacity    += ls[i] * ingredients[i][0]
        durability  += ls[i] * ingredients[i][1]
        flavor      += ls[i] * ingredients[i][2]
        texture     += ls[i] * ingredients[i][3]
        calories    += ls[i] * ingredients[i][4]
    capacity   = not_negative(capacity)
    durability = not_negative(durability)
    flavor     = not_negative(flavor)
    texture    = not_negative(texture)

    total = capacity * durability * flavor * texture
    max1 = max(max1, total)
    if calories == 500: max2 = max(max2, total)

    # --- Calculate next recipe ---
    i = last - 1
    carry = True
    while carry:
        limit = MAX
        if i > 0:
            for j in range(i): limit -= ls[i]
        ls[i] += 1
        if ls[i] > limit:
            ls[i] = 0
            if i > 0: i -= 1
            else:
                is_done = True
                break
        else:
            carry = False
        ls[last] = MAX
        for j in range(last): ls[last] -= ls[j]

# --- Results ---
print(f"1. Total score of the highest-scoring cookie: {max1:,d}")
print(f"2. Total score of the highest-scoring cookie: {max2:,d}")
